cpp/es2.1/date.cpp

```diff
@@ -0,0 +1,133 @@
+#include <iostream>
+
+using namespace std;
+
+class Date {
+private:
+    int dd;
+    int mm;
+    int yyyy;
+
+public:
+    // Costruttore di default
+    Date() {
+        dd = 1;
+        mm = 1;
+        yyyy = 1970;
+    }
+
+    // Costruttore con parametri
+    Date(int day, int month, int year) {
+        dd = day;
+        mm = month;
+        yyyy = year;
+    }
+
+    // Metodi getter
+    int day() const { return dd; }
+    int month() const { return mm; }
+    int year() const { return yyyy; }
+
+    // Operatore di output '<<'
+    friend ostream& operator<<(ostream& os, const Date& date) {
+        os << date.dd << "/" << date.mm << "/" << date.yyyy;
+        return os;
+    }
+
+    // Operatore di confronto '=='
+    friend bool operator==(const Date& one, const Date& other) {
+        return (one.dd == other.dd && one.mm == other.mm && one.yyyy == other.yyyy);
+    }
+
+    // Metodo per aggiungere giorni
+    void addDays(int num_d) {
+        // Calcolo numero giorni nel mese corrente
+        int giorniMese = getGiorniMese(mm, yyyy);
+
+        dd += num_d;
+
+        // Controllo incremento mese 
+        while (dd > giorniMese) {
+            dd -= giorniMese;
+            mm++;
+
+            // Controllo incremento anno 
+            if (mm > 12) {
+                mm = 1;
+                yyyy++;
+            }
+
+            giorniMese = getGiorniMese(mm, yyyy);
+        }
+    }
+
+    // Metodo per aggiungere mesi
+    void addMonths(int num_m) {
+        mm += num_m;
+
+        // Controllo incremento anno
+        while (mm > 12) {
+            mm -= 12;
+            yyyy++;
+        }
+
+        // Controllo lim numero giorni nel mese corrente
+        int giorniMese = getGiorniMese(mm, yyyy);
+        if (dd > giorniMese)
+            dd = giorniMese;
+    }
+
+    // Metodo per aggiungere anni
+    void addYears(int num_y) {
+        yyyy += num_y;
+
+        // Controllo lim numero giorni nel mese corrente
+        int giorniMese = getGiorniMese(mm, yyyy);
+        if (dd > giorniMese)
+            dd = giorniMese;
+    }
+
+private:
+    // Funzione per ottenere numero di giorni in un mese 
+    int getGiorniMese(int month, int year) const {
+        static const int giorniMese[] = {31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
+
+        // Gestione anni bisestili
+        if (month == 2 && (year % 4 == 0 && (year % 100 != 0 || year % 400 == 0)))
+            return 29;
+        
+        return giorniMese[month - 1];
+    }
+};
+
+int main() {
+    Date dataDefault;
+    cout << "Data di Default: " << dataDefault << endl;
+
+    Date dataCustom(20, 6, 2023);
+    cout << "Data Customizzata: " << dataCustom << endl;
```

Comment:
Come fai a testare che il costruttore `Date()` funzioni? :)
Devi testare anche quello in modo automatico. Quindi ti serve una cosa del tipo:
```c++
   if(dataCustom.day() == 20 && dataCustom.month() == 6 && dataCustom.year() == 2023 ) {
        cout << "[OK] Date(day, month, year)" << endl;
    } else {
        cout << "[FAIL] Date(day, month, year): " << dataCustom << " expected: (2023,6,20)" << endl;
    }
```